src/main/java/io/divolte/server/config/DivolteConfiguration.java

```diff
@@ -1,37 +1,131 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 @ParametersAreNonnullByDefault
+@MappingSourceSinkReferencesMustExist
+@SourceAndSinkNamesCannotCollide
+@OneSchemaPerSink
 public final class DivolteConfiguration {
-    @Valid public final ServerConfiguration server;
-    @Valid public final TrackingConfiguration tracking;
-    @Valid public final JavascriptConfiguration javascript;
-    @Valid public final IncomingRequestProcessorConfiguration incomingRequestProcessor;
-    @Valid public final KafkaFlusherConfiguration kafkaFlusher;
-    @Valid public final HdfsFlusherConfiguration hdfsFlusher;
+    @Valid public final GlobalConfiguration global;
+
+    // Mappings, sources and sinks are all keyed by their name.
+    @Valid public final ImmutableMap<String,MappingConfiguration> mappings;
+    @Valid public final ImmutableMap<String,SourceConfiguration> sources;
+    @Valid public final ImmutableMap<String,SinkConfiguration> sinks;
 
     @JsonCreator
-    private DivolteConfiguration(
-            final ServerConfiguration server,
-            final TrackingConfiguration tracking,
-            final JavascriptConfiguration javascript,
-            final IncomingRequestProcessorConfiguration incomingRequestProcessor,
-            final KafkaFlusherConfiguration kafkaFlusher,
-            final HdfsFlusherConfiguration hdfsFlusher) {
-        this.server = server;
-        this.tracking = tracking;
-        this.javascript = javascript;
-        this.incomingRequestProcessor = incomingRequestProcessor;
-        this.kafkaFlusher = kafkaFlusher;
-        this.hdfsFlusher = hdfsFlusher;
+    DivolteConfiguration(final GlobalConfiguration global,
+                         final Optional<ImmutableMap<String, SourceConfiguration>> sources,
+                         final Optional<ImmutableMap<String, SinkConfiguration>> sinks,
+                         final Optional<ImmutableMap<String,MappingConfiguration>> mappings) {
+        this.global = Objects.requireNonNull(global);
+        this.sources = sources.orElseGet(DivolteConfiguration::defaultSourceConfigurations);
+        this.sinks = sinks.orElseGet(DivolteConfiguration::defaultSinkConfigurations);
+        this.mappings = mappings.orElseGet(() -> defaultMappingConfigurations(this.sources.keySet(), this.sinks.keySet()));
+    }
+
+    public BrowserSourceConfiguration getBrowserSourceConfiguration(final String sourceName) {
+        final SourceConfiguration sourceConfiguration = sources.get(sourceName);
+        Objects.requireNonNull(sourceConfiguration, () -> "No source configuration with name: " + sourceName);
+        Preconditions.checkArgument(sourceConfiguration instanceof BrowserSourceConfiguration,
+                                    "Source configuration '%s' is not a browser source", sourceName);
+        return (BrowserSourceConfiguration)sourceConfiguration;
+    }
+
+    public MappingConfiguration getMappingConfiguration(final String mappingName) {
```

Comment:
Similar to above, I think the check for existence of the key is the responsibility of the calling code (Using `Map#containsKey(...)`).